crates/derive/src/stages/batch/batch_validator.rs

```diff
@@ -0,0 +1,622 @@
+    ///
+    /// ## Returns
+    /// - `true` if the origin is behind the parent origin.
+    fn origin_behind(&self, parent: &L2BlockInfo) -> bool {
+        self.prev.origin().map_or(true, |origin| origin.number < parent.l1_origin.number)
+    }
+
+    /// Updates the [BatchValidator]'s view of the L1 origin blocks.
+    ///
+    /// ## Takes
+    /// - `parent`: The parent block of the current batch.
+    ///
+    /// ## Returns
+    /// - `Ok(())` if the update was successful.
+    /// - `Err(PipelineError)` if the update failed.
+    pub(crate) fn update_origins(&mut self, parent: &L2BlockInfo) -> PipelineResult<()> {
+        // NOTE: The origin is used to determine if it's behind.
+        // It is the future origin that gets saved into the l1 blocks array.
+        // We always update the origin of this stage if it's not the same so
+        // after the update code runs, this is consistent.
+        let origin_behind = self.origin_behind(parent);
+
+        // Advance the origin if needed.
+        // The entire pipeline has the same origin.
+        // Batches prior to the l1 origin of the l2 safe head are not accepted.
+        if self.origin != self.prev.origin() {
+            self.origin = self.prev.origin();
+            if !origin_behind {
+                let origin = self.origin.as_ref().ok_or(PipelineError::MissingOrigin.crit())?;
+                self.l1_blocks.push(*origin);
+            } else {
+                // This is to handle the special case of startup.
+                // At startup, the batch validator is reset and includes the
+                // l1 origin. That is the only time when immediately after
+                // reset is called, the origin behind is false.
+                self.l1_blocks.clear();
+            }
+            debug!(
+                target: "batch-validator",
+                "Advancing batch validator origin to L1 block #{}.{}",
+                self.origin.map(|b| b.number).unwrap_or_default(),
+                origin_behind.then_some(" (origin behind)").unwrap_or_default()
+            );
+        }
+
+        // If the epoch is advanced, update the l1 blocks.
+        // Advancing epoch must be done after the pipeline successfully applies the entire span
+        // batch to the chain.
+        // Because the span batch can be reverted during processing the batch, then we must
+        // preserve existing l1 blocks to verify the epochs of the next candidate batch.
+        if !self.l1_blocks.is_empty() && parent.l1_origin.number > self.l1_blocks[0].number {
+            for (i, block) in self.l1_blocks.iter().enumerate() {
+                if parent.l1_origin.number == block.number {
+                    self.l1_blocks.drain(0..i);
+                    debug!(target: "batch-validator", "Advancing internal L1 epoch");
+                    break;
+                }
+            }
+            // If the origin of the parent block is not included, we must advance the origin.
```

Comment:
?

Reply:
This comment is correct - maybe better put above the loop, though?